Support tiered data storage

enigbe opened the pull request:
### What this PR does
In this PR we introduce `TierStore`, a three-tiered (`KVStore` + `KVStoreSync`) implementation that manages data across three distinct storage layers based on criticality.

#### Background
As we have moved towards supporting remote storage with `VssStore`, we need to recognize that not all data has the same storage requirements. Currently, all data goes to a single store which creates some problems:
- Remote storage adds latency on every read/write operation
- Network graph and scorer data is ephemeral and easily rebuild-able from the network
- Persisting ephemeral data to remote storage wastes bandwidth and adds unnecessary latency
- Users persisting to remote storage have no local disaster recovery option

This PR proposes tiered storage that provides granular control over where different data types are stored. The tiers include:
- Ephemeral store (network graph, scorer): optional fast local storage to avoid remote latency
- Primary store (channels, payments, wallet): preferably remote storage for critical data
- Backup store: optional (preferably local) backup for disaster recovery, with data written to (removed from) lazily to avoid blocking primary data operations

Additionally, we also permit the configuration of `Node` with tiered storage allowing callers to:
- set retry parameters,
- set backup and ephemeral stores, and to 
- build the `Node` with a primary store. 
These configuration options also extend to our foreign interface, allowing bindings target to build the `Node` with their own (`KVStore` + `KVStoreSync`) implementations. A sample Python implementation is provided and tested.

#### Concerns
1. Nested Retry Logic: `VssStore` has built-in retry logic. Wrapping it in `TierStore` creates nested retries.
2. Backup Queue Capacity: Currently hard-coded (100 in prod, 5 in tests). Ideally this should be configurable but there are concerns about exposing this to callers? What is a sensible default?
3. Data Consistency Between Primary and Backup: Backup operates asynchronously and may lag behind or miss operations if the queue overflows. Should we consider adding a background sync task to reconcile backup with primary? Expose metrics for backup lag/queue depth? Implement catch-up logic on startup?
4. Exposing `KVStore` to the FFI

#### Related Issues
- Addresses #408

tnull reviewed (commented):
Thanks for looking into this and excuse the delay here! 

I did a first pass, generally this looks already pretty good, but it is a huge PR and in some areas could be simplified. For instance, we should drop the generic `Retry`  logic as the concrete implementations would already implement that if they need it. Likewise, we shouldn't fallback to the backup for now as it's really only meant as disaster recovery (KVStore caching is out-of-scope for this PR now, even though we might want to explore that soon, too).  There is also no need to replicate the write-ordering locks in `TierStore`, *but* we'll need them in `ForeignKVStoreAdapter` (which might be mergeable with `DynStore`?). 

Generally, if you find opportunities to reduce the size of the changeset here it would be appreciated. It would also be cool if you could try to break up the PR into more feature commits, but feel free to leave as is if not.

tnull reviewed (commented):
Unfortunately this will need a substantial rebase now that #696 landed, sorry for that! 

Besides some tedious rebase work, we now have `DynStoreWrapper` which can be used on the ffi *and* the non-ffi side. I do wonder if we can actually just merge that wrapper with the `TierStore`, as well as the `ForeignKVStoreAdapter`/`DynStore`. Seems all these wrapper structs may not be needed and we could just get away with one trait and one wrapper struct, mostly?

enigbe commented on src/ffi/mod.rs:
```diff
@@ -45,3 +45,19 @@ pub fn maybe_try_convert_enum<T>(value: &T) -> Result<&T, crate::error::Error> {
+
+/// KVStore* wrapper
+#[macro_export]
+macro_rules! wrap_store {
```
This has been dropped.

enigbe commented on src/ffi/types.rs:
```diff
@@ -56,8 +60,330 @@ pub use crate::payment::store::{
+	) -> Pin<Box<dyn Future<Output = Result<(), lightning::io::Error>> + Send>> {
+		let inner = self.inner.clone();
+
+		let primary_namespace = primary_namespace.to_string();
```
This has now been added 1ba4831c195fd609bd0c496f58ead52eb51b27dd

enigbe commented on src/types.rs:
```diff
@@ -47,8 +47,11 @@ where
 /// A type alias for [`SyncAndAsyncKVStore`] with `Sync`/`Send` markers;
 pub type DynStore = dyn SyncAndAsyncKVStore + Sync + Send;
+#[cfg(feature = "uniffi")]
+pub(crate) use crate::DynStore;
```
I had initially used the same name and the feature gating became necessary to differentiate the types at call sites when uniffi was enabled.

enigbe commented on src/io/test_utils.rs:
```diff
@@ -352,3 +355,168 @@ pub(crate) fn do_test_store<K: KVStoreSync + Sync>(store_0: &K, store_1: &K) {
+	}
+}
+
+pub struct DelayedStore {
```
This has been removed here cb66b599da9b3db8b14f7b766af82ba8b604748a.
I couldn't think of any way better to test backup queue overflow not impacting primary writes. I agree that the boilerplate seems too much for just that single case.

tnull reviewed (commented):
Thanks for the rebase! Feel free to squash the fixups.

Here a a few initial comments, but this is still a huge PR. To make it a bit more manageable, I do wonder if we can split out the FFI-related changes to a follow-up and here only add the tiered store itself? We probably would make sure both land for the next release, but it might be a bit easier to review these steps separately.

TheBlueMatt commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,883 @@
+	///
+	/// The backup operates on a best-effort basis:
+	/// - Writes are queued asynchronously (non-blocking)
+	/// - No retry logic (We assume local store is unlikely to have transient failures).
```
Hmm, this is maybe something we should consider tweaking. Not necessarily in this PR, but this is a pretty substantial limitation.

TheBlueMatt commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,883 @@
+fn is_ephemeral_cached_key(pn: &str, sn: &str, key: &str) -> bool {
+	matches!(
+		(pn, sn, key),
+		(NETWORK_GRAPH_PERSISTENCE_PRIMARY_NAMESPACE, _, NETWORK_GRAPH_PERSISTENCE_KEY)
```
There's also `_SECONDARY_NAMESPACE`s for these constants, though they're all empty anyway.

TheBlueMatt reviewed (commented):
Just looked at `tier_store.rs`, I'll let @tnull handle the rest.

enigbe committed:
Implement tiered storage

This commit adds `TierStore`, a tiered `KVStore` implementation that
routes node persistence across three storage roles:

- a primary store for durable, authoritative data
- an optional backup store for a second durable copy of primary-backed data
- an optional ephemeral store for rebuildable cached data such as the
  network graph and scorer

TierStore routes ephemeral cache data to the ephemeral store when
configured, while durable data remains primary+backup. Reads and lists
do not consult the backup store during normal operation.

For primary+backup writes and removals, this implementation treats the
backup store as part of the persistence success path rather than as a
best-effort background mirror. Earlier designs used asynchronous backup
queueing to avoid blocking the primary path, but that weakens the
durability contract by allowing primary success to be reported before
backup persistence has completed. TierStore now issues primary and backup
operations together and only returns success once both complete.

This gives callers a clearer persistence guarantee when a backup store is
configured: acknowledged primary+backup mutations have been attempted
against both durable stores. The tradeoff is that dual-store operations
are not atomic across stores, so an error may still be returned after one
store has already been updated.

Additionally, adds unit coverage for the current contract, including:
- basic read/write/remove/list persistence
- routing of ephemeral data away from the primary store
- backup participation in the foreground success path for writes and removals

enigbe committed:
Integrate TierStore into NodeBuilder

Add native builder support for configuring ephemeral storage and a local
SQLite backup mirror.

Wrap the primary store in TierStore during node construction and create
configured secondary stores using dedicated SQLite database files.

Implement paginated listing through TierStore and update filesystem-backed
tests to use FilesystemStoreV2.

Add full-cycle integration coverage verifying durable backup mirroring.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,907 @@
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. You may not use this file except in
+// accordance with one or both of these licenses.
+#![allow(dead_code)] // TODO: Temporal warning silencer. Will be removed in later commit.
```
Let's not introduce this in the first place then. Warnings are fine to have for a commit.

tnull replied:
The fixup for this seems to be in the wrong commit.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,907 @@
+use std::future::Future;
+use std::sync::Arc;
+
+/// A 3-tiered [`KVStoreSync`] implementation that routes data across storage
```
`KVStore`, not `KVStoreSync`, right?

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,907 @@
+/// - an optional ephemeral store for non-critical, rebuildable cached data.
+///
+/// When a backup store is configured, writes and removals for primary-backed data
+/// are issued to the primary and backup stores concurrently and only succeed once
```
Hmm, is this really what we want? If the backup store is remote and we fail to persist, should we really abort? Doesn't make this our operations less reliable in practice instead of more, if we always fail on the weakest link?

Also, if we simply abort if one of the `write`s fails, how could we recover without inconsistencies? We don't have a good way to roll-back the write that went through, right? So we'd still have inconsistencies, so wouldn't it be preferable to not fail the write in the first place?

enigbe replied:
> Hmm, is this really what we want? If the backup store is remote and we fail to persist, should we really abort? Doesn't make this our operations less reliable in practice instead of more, if we always fail on the weakest link?

The current implementation follows Matt's [recommendation](https://github.com/lightningdevkit/ldk-node/pull/692#discussion_r2861738403), where we assume a simplified configuration with primary-remote and backup-local. In this case, we should have high confidence in local backup rarely failing. However, if it does, I do agree that we don't want to propagate its error, especially if primary-remote succeeds. I could modify it so that we log backup-local failures and only ever fail if primary fails.

> Also, if we simply abort if one of the writes fails, how could we recover without inconsistencies? We don't have a good way to roll-back the write that went through, right? So we'd still have inconsistencies, so wouldn't it be preferable to not fail the write in the first place?

I have thought about this and wonder if we should explore syncing by extending `DynStoreTrait` with `list_all_keys` and use LDK's existing `migrate_kv_store_data` to reconcile backup from primary at startup. This would require that primary and backup both impl `MigratableKVStore`. I'd like to know what you think about this approach.

Regarding primary-local, backup-remote, Matt suggested that I shelve the consideration for a later, possible follow up to this, if we can come up with a more robust solution than I initially implemented. Any pointers here will be deeply appreciated.

tnull replied:
> The current implementation follows Matt's [recommendation](https://github.com/lightningdevkit/ldk-node/pull/692#discussion_r2861738403), where we assume a simplified configuration with primary-remote and backup-local. In this case, we should have high confidence in local backup rarely failing. However, if it does, I do agree that we don't want to propagate its error, especially if primary-remote succeeds. I could modify it so that we log backup-local failures and only ever fail if primary fails.

Hum, maybe it might be best to  discuss all options we have offline @TheBlueMatt ? Especially if we also want to support remote backup (which is reasonable/important, IMO), failing if either operation fails seems not great? However, if we only want to support a local mirror for now, maybe we should just simplify everything, and make enabling the backup store a binary option that enables an SQLite-based backup store? As a side effect, the user could easily copy the database to recover?

> I have thought about this and wonder if we should explore syncing by extending `DynStoreTrait` with `list_all_keys` and use LDK's existing `migrate_kv_store_data` to reconcile backup from primary at startup. This would require that primary and backup both impl `MigratableKVStore`. I'd like to know what you think about this approach.

Yeah, likely using `MigratableKVStore` is the way to go, but maybe we should just leave this as a manual step rather than doing auto-recovery? We do already expose all necessary utilities for it, no?

> Regarding primary-local, backup-remote, Matt suggested that I shelve the consideration for a later, possible follow up to this, if we can come up with a more robust solution than I initially implemented. Any pointers here will be deeply appreciated.

Yeah, okay, if we really don't want to do remote backups for now, we should just keep a mirrored `SQLite` store at a configurable location, IMO, and not allow the user to configure anything else. I do however want to note that people might still try to misuse this, e.g., by pointing the storage location of the mirror to a remote file system (ask me how I know: I did the same thing for a while to keep my CLN node at the time backed up).

TheBlueMatt replied:
IMO if we want to move away from failing if either backend fails we need to have proper recovery logic in place. If we have a remote backup and writing to it fails, it otherwise wouldn't be obvious that restoring from the "live backup" is suddenly unsafe. Even making the write complete early suddenly makes using the "live backup" unsafe after a shutdown if we were mid-write.

It's great in theory to provide an optimistic live backup, but absent a way to actually use it safely in really not convinced it's a great idea to ship it.

tnull replied:
Right, so IMO the short term solution is to hardcode backups to local SQLite, and treat write failures the same as persistence failures to the primary store.

Ofc, that won't safe us entirely from the two stores getting out of sync as we'd still won't have rollbacks in place in case one write succeeds and the other doesn't. Question is if that means that we can at least parallelize instead of sequencing the writes.

TheBlueMatt replied:
That should be fine, though, as long as async writes aren't completed until both complete. IOW, it should always be the case that we can restart in either state when there's a pending uncompleted write when we stop. I don't see any reason to not parallelize the writes.

TheBlueMatt replied:
I guess we didn't strictly have any kind of guarantees around it being allowed for write A to be incomplete, then on restart missing, then later completed (in case we switched the "primary store" after a later restart). At least in the `lightning` crate I think that's fine - for monitors it'd only matter in cases where another update happened, but that would imply overwriting the previous dangling write to A. Similar is obviously true for nearly everything else, where we're constantly overwriting the latest state.

TheBlueMatt replied:
I guess we should document this on `KVStore` so that its a formal part of the API.

enigbe replied:
> Yeah, okay, if we really don't want to do remote backups for now, we should just keep a mirrored SQLite store at a configurable location, IMO, and not allow the user to configure anything else

@tnull: This has been implemented now, alongside parallelizing async writes and ensuring writes aren't completed until both succeed. I've reverted all changes related to the async backup retry queue and will open an RFC to propose the implementation and seek feedback before opening a PR.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,907 @@
+/// Reads and lists do not consult the backup store during normal operation.
+/// Ephemeral data is read from and written to the ephemeral store when configured.
+///
+/// Note that dual-store writes and removals are not atomic across the primary and
```
Okay, as mention above, I noted that: but what can I do now if a write failed? How can the user actually take action to avoid this or recover from it?

enigbe replied:
With the proposed change (only fail if primary fails), the user never needs to act on backup failures, we can just log them and reconcile at startup. Primary failure handling remains as is.

enigbe replied:
I have spent some time rethinking this and considered the option of allowing users to configure how they'd like to backup data, i.e. either best effort or semi-synchronously. In the latter case, in addition to the synchronous backup store, I have added an asynchronous retry queue backed by a local `FilesystemStoreV2` where we try to durably persist our backup intent. 

We offer guarantees in semi-sync mode based on the durable persistence of primary write and backup intent, surfacing a write failure to the caller only if the backup intent cannot be persisted to the local retry queue, even though the primary write may have already succeeded. I believe this is better than the alternative where we silently lose backup coverage without the caller knowing.

This does create a scenario where the backup store can drift behind the primary: if the node crashes after the primary write succeeds but before backup intent to retry queue is persisted, the backup will be missing those operations. The backup will be internally consistent but stale, and drift between the stores will form as the retry queue will be unaware of backup intent(s) that didn't persist. I'd like to understand whether this guarantee is sufficient for the restore use case, or whether we'd also need active reconciliation between the stores.

This approach was inspired by how DDIA describes semi-synchronous replication in single-leader replication systems. I also think it addresses some of the concerns @TheBlueMatt raised about primary-local and backup-remote configurations.

Would be really grateful to get both your thoughts on the approach taken.

tnull replied:
Thanks for thinking through this, however, it would be great if we could move any more advanced scheme like a backup queue to a follow-up PR as that might need some more review time and we should aim to land this PR (and the bindings follow-up) soon. So, as mentioned above, how about we just make the backup store an SQLite store with configurable location, requiring both writes to continue. Then, in a follow up we can explore the queuing concept. 

For one that gives us more flexibility with regards to when this follow-up lands, but it would also allow us to further think through, e.g., how to deal with the overlap with a next-gen `VssStore`, which should also start using a local read cache, *and* how all of this would interact with the generally keeping only a LRU page cache of the (payment) stores in memory and reading the entries from the stores on demand (which we also want to finally do soon).

tnull commented on src/builder.rs:
```diff
@@ -614,6 +633,36 @@ impl NodeBuilder {
+	/// primary and backup stores complete successfully.
+	///
+	/// If not set, durable data will be stored only in the primary store.
+	#[allow(dead_code)] // Used by subsequent FFI/test integration commits.
```
Why is this allowing `dead_code`?

enigbe replied:
I added this to silence warnings as I didn't think warnings were permissible in intermediate commits. Since you've clarified that they are, I can remove them from intermediate commits. However, the annotations on the final commit are still needed (the code isn't used except in #871), so CI will fail without them.

tnull commented on src/builder.rs:
```diff
@@ -614,6 +633,36 @@ impl NodeBuilder {
 		self
 	}
 
+	/// Configures the backup store for local disaster recovery.
```
So here we set the backup store, but how do we envision the restore to work? Should that be part of the `recovery_mode`?

enigbe replied:
There are two approaches we can take here. For the first, have both primary and backup concrete stores implement `MigratableKVStore` and have the user call `migrate_kv_store_data` before building. It's simple but not ideal as it's not part of any existing node or builder APIs, and would require explicit documentation.

Alternatively, we can add a `restore_from_backup(backup)` method on `NodeBuilder` and have the migration/restoration of data from backup (source) to primary (target) happen inside build. This requires adding `list_all_keys` to `DynStoreTrait`, and `MigratableKVStore` for both stores so the migration can work through the type-erased `Arc<DynStore>` layer.

For the second approach, we could also refactor `recovery_mode` from a bool into a struct:
```rust
pub struct RecoveryMode {
    pub wallet: bool, // resync on-chain wallet from genesis
    pub backup: bool, // restore persisted data from backup store
}
```
This keeps both recovery concerns under one concept while remaining independent. A user restoring from backup may not need a full wallet resync, and vice versa.

enigbe replied:
On the restore side, I went with the second approach from my earlier comment `restore_from_backup()` on the builder, with restoration happening inside build before normal startup reads. I refactored `recovery_mode` from a bool into a `RecoveryConfig` struct that keeps wallet recovery and backup restore independent. The restore itself enumerates the backup store via `list_all_keys`, filters to a known durable key inventory, checks that the primary store is empty, and copies only matching entries. This keeps restore policy explicit rather than blindly migrating everything from the backup store.

tnull commented on tests/common/mod.rs:
```diff
@@ -457,9 +457,7 @@ impl Default for TestConfig {
-		let $builder = Builder::from_config($config.clone());
-		#[cfg(not(feature = "uniffi"))]
+		#[allow(unused_mut)]
 		let mut $builder = Builder::from_config($config.clone());
```
I'm confused: why do we suddenly need `mut` under `uniffi`, especially given we don't add any bindings in this PR now?

enigbe replied:
I had previously introduced a test helper `setup_node_with_builder` that took a closure that required the builder to be mutable. This affected all tests (including uniffi gated ones) that utilized the helper as wrapped in `setup_node`. In hindsight, it was an unnecessary refactor. I've addressed the removal in https://github.com/lightningdevkit/ldk-node/pull/692/commits/1ef68f2147e830934263e4572aac022dec3b9e8e

tnull commented on src/builder.rs:
```diff
@@ -625,6 +655,42 @@ impl NodeBuilder {
+	/// can be rebuilt if lost.
+	///
+	/// If not set, non-critical data will be stored in the primary store.
+	pub fn set_ephemeral_store(&mut self, ephemeral_store: Arc<DynStore>) -> &mut Self {
```
I think the ephemeral store should always be a simple SQLite. Note that `DynStore` might currently not even be constructed outside of the crate right now.

enigbe replied:
Noted. I thought we'd leave the store type decision to the users and made `DynStore` public for this reason, but as stated [here](https://github.com/lightningdevkit/ldk-node/pull/692#discussion_r3333738995), the decision has been reverted for SQLite.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,905 @@
+		&self, primary_namespace: &str, secondary_namespace: &str, key: &str, buf: Vec<u8>,
+	) -> io::Result<()> {
+		if let Some(backup_store) = self.backup_store.as_ref() {
+			let primary_res = KVStoreSync::write(
```
Codex:

>   - [P1] Serialize primary and backup writes per key — /home/tnull/workspace/ldk-node/src/io/tier_store.rs:319-327
>     When two operations for the same key run at the same time, this method writes to the primary and backup without a shared per-key lock. One call can update the primary, a second can update both stores, and then the first can update the backup, so both calls return Ok(()) while the backup
>     contains an older value than the primary. Please serialize each primary+backup write/remove pair per key in both the sync and async paths.


Hmm, maybe we need some kind of key-level locks afterall? WDYT?

enigbe replied:
The intuition is correct here. A possible DAG without per-key lock is:
[w<sub>1_prim</sub>] -> [w<sub>2_prim</sub>] -> [w<sub>2_bck</sub>] -> [w<sub>1_bck</sub>] with the backup ending up with stale data from an out-of-order write.

I've re-added the per-key locks and also added `clean_locks` (following the same pattern as `VssStore`) to prune entries when no in-flight operations remain for a key.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,905 @@
+		}
+	}
+
+	fn ephemeral_store(
```
This method is super confusing, as it makes the callsite look as if we always delegate to the ephemeral store if it's set. Can we rather inline this, ie., have the callsites use `is_ephemeral_cached_key` to decide and then delegate to `self.ephemeral_store` when needed?

enigbe replied:
This has been updated.

tnull commented on src/builder.rs:
```diff
@@ -1109,6 +1212,38 @@ impl ArcedNodeBuilder {
+	/// If not set, durable data will be stored only in the primary store.
+	///
+	/// [`SQLITE_BACKUP_DB_FILE_NAME`]: crate::io::sqlite_store::SQLITE_BACKUP_DB_FILE_NAME
+	pub fn set_backup_storage_dir_path(&self, backup_storage_dir_path: String) {
```
I guess adjusting `ArcedNodeBuilder` should be part of the `uniffi` follow-up?

enigbe replied:
Yes. Moving this to the #871

tnull commented on src/types.rs:
```diff
@@ -57,7 +57,7 @@ where
 }
 
-pub(crate) trait DynStoreTrait: Send + Sync {
+pub trait DynStoreTrait: Send + Sync {
```
Why is this `pub` now?

enigbe replied:
This was made public so users could create their own ephemeral `DynStore` but since you've suggested that it be a SQLite store, I'll revert the change here.

tnull reviewed (commented):
Excuse the delay here! Mostly looks good, just a few comments.

enigbe commented:
Hi @tnull 
Thanks for the review again. I've addressed the latest feedback in a series of fixup commits.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,702 @@
+			primary_namespace,
+			secondary_namespace,
+			key,
+			buf.clone(),
```
Hmm, can we avoid cloning if no backup store is set?

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,1270 @@
+		Arc::clone(&outer_lock.entry(locking_key).or_default())
+	}
+
+	async fn execute_locked_write<
```
I think I have to eat my words here: it seems we have to implement key-level versioning for `TierStore` too as the returned write/remove futures are not immediately polled.

Codex:

>   - P1 /home/tnull/workspace/ldk-node-pr-692/src/io/tier_store.rs:100: TierStore::write/remove delays delegation until the returned future is polled, so it
>     loses the existing stores’ “latest call wins” ordering. Example: create old = write(old), then new = write(new), await new, then await old; TierStore will
>     write old last. Existing SqliteStore assigns the version at call time and skips stale writes (/home/tnull/workspace/ldk-node-pr-692/src/io/sqlite_store/
>     mod.rs:79, /home/tnull/workspace/ldk-node-pr-692/src/io/sqlite_store/mod.rs:603). This can corrupt async channel monitor persistence by letting an older
>     pending update overwrite or remove newer state. TierStore needs its own call-time per-key versioning before returning the future.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,702 @@
+		{
+			Ok(data) => Ok(data),
+			Err(e) => {
+				log_error!(
```
Codex:

>  - P3 /home/tnull/workspace/ldk-node-pr-692/src/io/tier_store.rs:197: read_primary logs every read error as ERROR, including normal NotFound startup paths.
>     Since the builder now wraps every store in TierStore (/home/tnull/workspace/ldk-node-pr-692/src/builder.rs:897), fresh nodes will log expected missing
>     node_metrics, graph, scorer, event queue, peers, etc. as errors before callers handle them as defaults. The wrapper should avoid error-level logging for
>     NotFound and generally let callers decide.

enigbe replied:
Done.

tnull reviewed (commented):
Feel free to squash fixups!

Unfortunately, `cargo test --features uniffi` currently fails to compile currently. I think we might also need to revert to implement the same key-level versioning pattern as we have in other stores afterall. Excuse the back-and-forth.

enigbe commented:
Hi @tnull,

I've addressed the most recent feedback in new fix-up commits.

tnull commented on src/io/tier_store.rs:
```diff
@@ -4,7 +4,6 @@
 // http://www.apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
 // http://opensource.org/licenses/MIT>, at your option. You may not use this file except in
 // accordance with one or both of these licenses.
-#![allow(dead_code)] // TODO: Temporal warning silencer. Will be removed in later commit.
```
Please avoid unnecessarily adding things that you'll end up removing in later commits.

enigbe replied:
I've dropped the commit but CI will fail for `uniffi` builds.

tnull replied:
Hmm, my comment was mostly about the fact that, AFAIR, you first added the `dead_code` attribute, then removed it in a later commit. But in general it's also better to avoid such 'general' `dead_code` attributes as they always have the risk of getting missed and eventually will lead to dead code lingering in the codebase. 

For now, failing CI is of course not an option, so you could either tag the respective methods with `#[cfg(not(feature = "uniffi"))]` or do `#[cfg_attr(feature = "uniffi", dead_code)]` to at least gate the exception on the `uniffi` feature.

enigbe replied:
Sure.  I'll tag the methods of concern.

tnull commented on tests/common/mod.rs:
```diff
@@ -601,10 +602,6 @@ pub(crate) fn setup_node(chain_source: &TestChainSource, config: TestConfig) ->
 		},
 	};
 
-	if config.recovery_mode {
```
Why was this removed?

enigbe replied:
Possibly happened during a rebase.

tnull commented on tests/common/mod.rs:
```diff
@@ -532,6 +532,7 @@ pub(crate) fn setup_two_nodes_with_store(
 
 pub(crate) fn setup_node(chain_source: &TestChainSource, config: TestConfig) -> TestNode {
 	setup_builder!(builder, config.node_config);
+
```
Please avoid all these unnecessary/unrelated whitespace changes.

tnull commented on src/builder.rs:
```diff
@@ -842,6 +894,39 @@ impl NodeBuilder {
+					log_error!(logger, "Failed to setup backup SQLite store: {}", e);
+					BuildError::KVStoreSetupFailed
+				})?;
+				let backup_store: Arc<DynStore> = Arc::new(DynStoreWrapper(backup_store));
```
Do we need to detect whether the backup store is empty and use `MigratableKVStore` to backfill the backup store before we continue operating if this is enabled on a pre-existing node? If it's disabled and re-enabled at a later date, how could we detect that the backup is stale and needs to be resilvered?

We probably need to find solutions here, but maybe they should also happen in a follow-up PR, hopefully still landing for the next release?

enigbe replied:
These are good questions I didn't consider but I will in the follow up to this PR.

tnull replied:
Can you open an issue for backfilling so we don't forget to look into it?

enigbe replied:
Yes. Doing so now.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,816 @@
+
+		if is_ephemeral_cached_key(&primary_namespace, &secondary_namespace, &key) {
+			if let Some(eph_store) = self.ephemeral_store.as_ref() {
+				let res = KVStore::write(
```
Codex:

>   - [P2] Preserve call ordering for ephemeral writes — /home/tnull/workspace/ldk-node/src/io/tier_store.rs:403-411
>     With an ephemeral store configured, network graph/scorer writes bypass execute_locked_write; because the underlying store assigns its write version only
>     when this future is polled, a future created earlier but awaited after a newer one can overwrite the newer ephemeral cache. This is the same out-of-order
>     scenario covered for primary-backed writes, and the remove branch can similarly delete a newer write, so ephemeral writes/removes should use the same
>     version check.

tnull reviewed (commented):
Thanks! Feel free to squash fixups into the respective commit and drop the DROPME commit.

tnull reviewed (commented):
Unfortunately this needs yet another rebase by now.

enigbe committed:
fixup! Implement tiered storage

Preserve call-time ordering for ephemeral writes and removes by routing them
through the same versioned lock path as primary-backed mutations.

Add regression coverage for stale ephemeral writes and removes.

enigbe commented:
Hi @tnull 
This has been rebased and is ready for another look. CI is failing but that's expected because the methods to configure the backup and ephemeral stores are never used.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,1050 @@
+			"list",
+		)?;
+
+		match (primary_namespace.as_str(), secondary_namespace.as_str()) {
```
Codex:

>   - [P2] Merge root namespace listings across tiers — /home/tnull/workspace/ldk-node/src/io/tier_store.rs:497-503
    With an ephemeral store configured, this namespace-only match catches the empty root namespace because LDK's network graph/scorer namespaces are ("", ""). Durable root keys written through this same TierStore to the primary tier, such as manager, events, peers, node_metrics, and
    output_sweeper, then cannot be discovered by list("", ""), even though direct reads still succeed; the duplicated paginated branch has the same problem.

enigbe replied:
Merging paginated listings is not as straightforward as it is for `list`. Between the primary and ephemeral stores, there is no cross-store shared sort order via which we can guarantee paginated listings return in line with the expected creation order (most-recently created first) for `list_paginated`. 

`PageToken` and the underlying creation-order mechanism are deliberately opaque per the `PaginatedKVStore` trait, and `TierStore` is written against that abstraction, not concrete store types, so it has no portable way to extract or compare an ordering value across two backends even where they might happen to use a comparable mechanism internally.

`TierStore` can guarantee correct within-store ordering for primary and for ephemeral separately, but not a true globally-sorted merge across them (as currently implemented, `TierStore` doesn't even query ephemeral's paginated API for this, since a correctly-ordered overlay still couldn't be merged into primary's pages). This is why the implementation (https://github.com/lightningdevkit/ldk-node/pull/692/commits/81e65cf1334733bfbe03206294892dd652509d19) pages through primary however many calls it takes, and only appends the ephemeral overlay once primary reports no more pages, instead of trying to interleave creation time.

Alternatively, we could consider tracking creation order on `TierStore`, but I'd prefer to get your feedback on this first.

benthecarman replied:
Maybe the best way to solve this is just change the `PaginatedKvStore` to have the page token just be the last element in the sequence.

tnull reviewed (commented):
... and unfortunately needs yet another minor rebase (sorry, a bunch of PRs are going in right now). 

Otherwise we should be close to landing, mod the open comment above.

enigbe committed:
fixup! Implement tiered storage

Previously, the implementations for `list` and paginated listing on `TierStore`
contained a bug where, if the ephemeral store was configured, these operations
could return only keys from the ephemeral tier for shared root namespaces such
as ("", ""). This meant durable keys saved to the primary store could be ignored
whenever their namespace was shared with ephemeral-cached keys.

While the solution to `list` seemed straightforward, the one for paginated
listing raised questions about the potential difference in store types between
the primary and ephemeral stores, and the conflict that would arise from them
having independent pagination orders and tokens. `TierStore`, as currently
designed, does not maintain a cross-store cursor, so we rely on a simple overlay
strategy: paginate through primary first, then append ephemeral keys once primary
is on its terminal page.

This does mean the resulting paginated listing is not a strict cross-store
creation-order merge: ephemeral keys are returned after primary pagination is
exhausted, regardless of their creation order in the ephemeral store. We accept
that tradeoff because `TierStore` has no shared creation order across backing
stores, and the overlay is bounded to the current ephemeral-cached key set.

This commit allows us to merge listings across `TierStore`'s primary and
ephemeral tiers by treating the latter as an overlay for ephemeral-cached keys
only. We filter stale primary copies of ephemeral-cached keys, append live
ephemeral copies once primary pagination is complete, and only consult the
ephemeral store for namespaces that can contain ephemeral-cached keys.

enigbe commented:
Rebased now and addressed the pending comment.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,1464 @@
+		// full-sized terminal page can make it up to `MAX_EPHEMERAL_CACHED_KEYS_PER_NAMESPACE`
+		// entries larger than the nominal page size. Both effects are bounded by that constant.
+		let append_live = response.next_page_token.is_none();
+		self.apply_ephemeral_overlay(
```
I we shouldn't probably use this impl for any usecase that wouldn't require primary store access. Rather than adding lots of complexity around an implementation that we know is broken, let's just list from the primary store here, leave a TODO and open an issue that we still have to resolve this. Then we can probably move forward with this PR finally, also so you can rebase the Uniffi work.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,1464 @@
+/// and can't correctly report `next_page_token` for a partial overlay.
+const MAX_EPHEMERAL_CACHED_KEYS_PER_NAMESPACE: usize = 2;
+
+fn is_ephemeral_cached_key(pn: &str, sn: &str, key: &str) -> bool {
```
This also needs to include `external_pathfinding_scores_cache`.

tnull commented on src/io/tier_store.rs:
```diff
@@ -0,0 +1,1464 @@
+			if version <= *last_written_version {
+				Ok(())
+			} else {
+				let res = callback().await;
```
Codex:

>  - [P1] Failed backup writes can roll back newer primary state. /home/tnull/worktrees/ldk-node/pr-692-review-20260714/src/io/tier_store.rs:366 advances only when both stores succeed. If a newer write updates primary but backup fails, an older future awaited afterward can overwrite primary
>     with stale data, violating KVStore write ordering.

tnull reviewed (commented):
Excuse the delay, this need yet another rebase by now. 

Please squash fixups and also address the first two comments (simplify the known-broken paginated list and add `external_pathfinding_scores_cache`). The rest can happen in a follow-up (and we need to figure out a way to fix the pagination at some point, but that will likely be another PR entirely).

tnull commented:
Moving this off the milestone as it seems abandoned and we hit some issues around merging the pagination logic across different backends. Will need to revisit at some point.

enigbe commented:
> Moving this off the milestone as it seems abandoned and we hit some issues around merging the pagination logic across different backends. Will need to revisit at some point.

Whoa! This hasn’t been abandoned. I took some time away to work on a research project and have now returned. I’ve pushed fixup commits addressing the latest review feedback. Could you please reconsider adding it back to the current milestone?

tnull commented:
> Whoa! This hasn’t been abandoned. I took some time away to work on a research project and have now returned. I’ve pushed fixup commits addressing the latest review feedback. Could you please reconsider adding it back to the current milestone?

Oh, sorry, I didn't mean to be too hasty. That said, it still seems we don't have worked out a great way to account for the pagination issue, and given that after this one the follow-up PR would also need to land in the same release, I doubt they will make it both in time (as we will attempt to finally ship v0.8 in the next ~2 weeks).
